### PR TITLE
Dynamic programming memoization#1833

### LINKED ISSUE #1833: A beginner's guide to memoization and Dynamic Programming in Python
### Proposed title of article
A beginner's guide to memoization and Dynamic Programming in Python

### Introduction paragraph (2-3 paragraphs):
Memoization is the process of saving the results of a function to a cache to improve its performance. It is the technique where we ensure that the function does not run for the same inputs more than once; instead, we return a cached result to reduce the running time of an algorithm.

Dynamic programming is an algorithm technique where we break down a problem into simpler, smaller subproblems, find the solutions to these sub-problems then use these solutions to provide an optimal solution to the original problem.
In this article we are going to introduce the reader to these two concepts using relevant examples then we are going to use them to solve a common optimization problem.

### Key takeaways: 
1. What is Dynamic programming.
2.  What is memoization.
3. How to use memoization and DP in python
4. How to solve the 0-1 Knapsack problem  using DP and memoization in python

### References: 
N/A

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by elkiplangat**
### Article checklist

**ATTENTION: In your PR - add the ISSUE (topic form) # that this PR corresponds to. You can do this by also adding a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.**
      
      
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**If you have not already please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve our overall technical writing.**

#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**

This fixes #1833

### CODE REVIEW

**Comment by srishilesh** [comment]
@elkiplangat 

Welcome to EngEd community!

Based on your commits, I notice 2 different `index.md` files. Please ensure you only upload one `index.md` file.

Also, make sure to upload a hero image of dimension 600x400. You can use Unsplash to download images, and resize it using Pixlr.

Add author's information - author bio and headshot image under `/content/authors`.

Please refer the contribution guide to understand better.

Thank you.

**Comment by elkiplangat** [comment]
@srishilesh fixed. I've added a hero image and fixed the double index.md issue. MY author information is already in the /authors directory.

File: content/articles/memoization-and-dp-in-python-beginners-guide/index.md [hunk 1]
```diff
@@ -0,0 +1,122 @@
+###A Beginner's guide to memoization and Dynamic Programming in Python
+
+as we saw in [this]('https://www.section.io/engineering-education/introduction-to-dynamic-programming/) article, Dynamic programming is the algorithm technique where we break down a problem into smaller, simpler subproblems then find the solutions to these smaller subproblems and use these solutions to find an optimal solution to the original problem.
```
Comment (srishilesh) [inline content/articles/memoization-and-dp-in-python-beginners-guide/index.md:3]:
```suggestion
As we saw in [this](/engineering-education/introduction-to-dynamic-programming/) article, Dynamic programming is the algorithm technique where we break down a problem into smaller, simpler subproblems then find the solutions to these smaller subproblems and use these solutions to find an optimal solution to the original problem.
```

File: content/articles/memoization-and-dp-in-python-beginners-guide/index.md [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+[Memoization](#memoization)
+
+[The 0-1 knapsack problem](#knapsack)
+
+[Python Solution to the 0-1 knapsack problem](#implementation)
+
+[How to apply memoization to the solution](#solution)
+
+###prerequisites<a name='prerequisites'></a>
```
Comment (srishilesh) [inline content/articles/memoization-and-dp-in-python-beginners-guide/index.md:17]: Remove these `<a></a>` anchor tags for every headings

File: content/articles/memoization-and-dp-in-python-beginners-guide/index.md [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+            return val
+        if weights[n-1]<=W:
+            memodict[key] = max(values[n-1]+knapsack(W-weights[n-1], weights, values, n-1), knapsack(W, weights, values, n-1))
+        elif weights[n-1]>W:
+            memodict[key] = knapsack(W, weights, values, n-1)
+
+        return memodict[key]
+
+```
+#Complexity Analysis
```
Comment (srishilesh) [inline content/articles/memoization-and-dp-in-python-beginners-guide/index.md:110]: `#` -> `###`

File: content/articles/memoization-and-dp-in-python-beginners-guide/index.md [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+For the unmemoized version, the time complexity is O(2^n) because there are redundant calculations we are making.
+For the memoized version, the time complexity is O(n*W) that is the table size by the capacity of the knapsack.
+
+#Conclusion
```
Comment (srishilesh) [inline content/articles/memoization-and-dp-in-python-beginners-guide/index.md:114]: `#` -> `###`

File: content/articles/memoization-and-dp-in-python-beginners-guide/index.md [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+Memoization is a useful trick to have under your belt when solving a dynamic programming problem, we've seen how we can use it to decrease the time complexity
+from 2^n to n*W.
+
+#Resources
```
Comment (srishilesh) [inline content/articles/memoization-and-dp-in-python-beginners-guide/index.md:118]: `#` -> `###`

File: content/articles/memoization-and-dp-in-python-beginners-guide/index.md [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+    array is an array of n+1 numbers
+    '''
+    if n<=1:
+        array[n]=n
+    if array[n] is None:
+        array[n] = mem_fib(n-1, array)+mem_fib(n-2, array)
+    return array[n]
+```
+In the memoized version you can see that we started from the values n-1 and n-2 going downwards.
+Now that you know what memoization is, let's use it to solve a common programming problem.
```
Comment (srishilesh) [inline content/articles/memoization-and-dp-in-python-beginners-guide/index.md:47]: Don't just blindly state facts or copy stuff from other articles.

Instead walk the reader through the content. Explain them step-by-step. Explain how it works. Why we memoize it? What if we do not? Stack trace the program for sample input

File: content/articles/memoization-and-dp-in-python-beginners-guide/index.md [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+###The 0-1 Knapsack problem<a name='knapsack'></a>
+The 0-1 knapsack problem is an optimization problem where you are given a set of items, their weights and values, and a knapsack. Your task is to determine the combination of items that are going to give maximum value without reaching the weight limit of your knapsack.
+For every item in the list, we can either include it in the knapsack or discard it(This is where the 0-1 name comes from).
+In the coding implementation of this problem, you are given two arrays, one is the weights array and the other is the values array. You are also given a weight W which is the maximum weight the knapsack can carry.
+From the definition alone, we can see that this problem can be solved using Dynamic Programming. This is because it has overlapping subproblems that when solved lead towards solving the whole problem.
```
Comment (srishilesh) [inline content/articles/memoization-and-dp-in-python-beginners-guide/index.md:52]: `From the definition alone, we can see that this problem can be solved using Dynamic Programming.`

Beginners might not be aware of DP or other programming stuff. You need to break them down.

Explain what DP is, why we use it? How is it related with DP? Why not brute force?

File: content/articles/memoization-and-dp-in-python-beginners-guide/index.md (hunk 3 above)
Comment (srishilesh) [inline content/articles/memoization-and-dp-in-python-beginners-guide/index.md:110]: This is a huge topic.

I can't believe you just covered them in 2 lines. How did you derive this time or space complexity. Explain step by step.

Before all, first explain what complexity analysis is.

**Review by srishilesh — CHANGES_REQUESTED** [review]
@elkiplangat 

I've suggested some major changes.

Also, perform a grammar check. I found many errors in capitalization of letters, and punctuation errors. Please correct them.

This article is too short to explain this vast concept. Re-read your article from the perspective of a beginner to see if you can understand.

Explore more on what makes your article different from the already existing articles on GFG or Hackerearth. And, see how you can improve this better. Though, most of the content in your article is already found on the internet, I guess you could explain them in a much better way.

Thanks.

**Comment by srishilesh** [comment]
@elkiplangat 

Any updates?

**Comment by elkiplangat** [comment]
@srishilesh you suggested some major changes, I'm gonna work on them during the weekend when I have more time.

**Comment by srishilesh** [comment]
@elkiplangat 

No worries. Take your own time.

**Comment by hectorkambow** [comment]
plagiarism check done

**Comment by hectorkambow** [comment]
@elkiplangat Any update on this ? 

If you need more time to apply the changes no worries what so ever when can close the PR - and reopen when ready. 👍 
cc @srishilesh

**Comment by hectorkambow** [comment]
Just closing to make space in the queue - we can re-open it when the article is ready to be published.

**Pull request closed without merging**